operators/src/pro/cache/tile_cache.rs

```diff
@@ -29,16 +30,425 @@ pub struct TileCacheBackend {
+    }
+
+    fn landing_zone_and_size_mut(
+        &mut self,
+        key: &CanonicOperatorName,
+    ) -> Option<LandingZoneSizeRef> {
+        self.operator_caches
+            .get_mut(key)
+            .map(|cache| (&mut cache.landing_zone, &mut self.landing_zone_size))
+    }
+
+    pub fn query_cache<T>(
+        &mut self,
+        key: &CanonicOperatorName,
+        query: &RasterQueryRectangle,
+    ) -> Option<CacheTileStream<T>>
+    where
+        T: Pixel + Cachable,
+    {
+        let (cache, _cache_size, cache_lru) = self.cache_and_size_and_lru_mut(key)?;
+
+        let mut expired_ids = Vec::new();
+        let x = cache.iter().find(|&(id, r)| {
+            if r.is_expired() {
+                expired_ids.push(*id);
+                return false;
+            }
+            r.matches(query)
+        });
+
+        let res = if let Some((id, entry)) = x {
+            // set as most recently used
+            cache_lru.promote(id);
+            T::stream(entry.tiles.tile_stream(query))
+        } else {
+            None
+        };
+
+        // remove expired entries
+        self.remove_querys_from_cache_and_lru(key, expired_ids.as_slice());
+
+        res
+    }
+
+    pub fn insert_query<T: Pixel + Cachable>(
+        &mut self,
+        key: &CanonicOperatorName,
+        query: &RasterQueryRectangle,
+    ) -> Result<QueryId, CacheError> {
+        // if there is no cache for this operator graph, we create one. However, it is not clear yet, if we will actually cache anything.
+        if !self.operator_caches.contains_key(key) {
+            self.operator_caches.insert(key.clone(), Default::default());
+            // self.cache_size.try_add_element_bytes(&key)?; // TODO: can we delete this at a later point?
+        };
+
+        let query_id = QueryId::new();
+
+        let landing_zone_entry = LandingZoneEntry {
+            query: *query,
+            tiles: T::create_active_query_tiles(),
+        };
+
+        let (landing_zone, landing_zone_size) = self
+            .landing_zone_and_size_mut(key)
+            .expect("Landing zone must must exist since we just created it");
+
+        // This gets or creates the cache entry for the operator graph. If it already exists, we don't need to add the size of the key again (see if case above).
+        landing_zone_size.try_add_element_bytes(&query_id)?;
+        landing_zone_size.try_add_element_bytes(&landing_zone_entry)?;
+        let old_entry = landing_zone.insert(query_id, landing_zone_entry);
+
+        if let Some(old_entry) = old_entry {
+            debug!("Replacing old entry in landing zone!");
+            landing_zone_size.remove_element_bytes(&old_entry); // Since the old entry already was in the landing zone, we need to remove it from the size.
+        }
+
+        Ok(query_id)
+    }
+
+    fn remove_query_from_landing_zone(&mut self, key: &CanonicOperatorName, query_id: &QueryId) {
```

Comment:
would it make sense to delete the whole `operator_caches` entry if the entry would be empty (no landing zone entries and no cache entries)?

Reply:
d06117e815802be648856c9b0850248e4eca1fcc

Reply:
also [7ed4001](https://github.com/geo-engine/geoengine/pull/819/commits/7ed40014e78b19ba5ecd79fd62570a35caa30ffe) and [38bdd50](https://github.com/geo-engine/geoengine/pull/819/commits/38bdd50817d933563bd94bdbff518a1d93b5558a)